File: src/mcp/src/analysis.ts
import { Evaluation, ClassifiedMove, GeneratedMove, GameState, MoveClassification } from './types.js';
import { generateNextStates, HashCalc } from './shogi.js';
import { LookupDatabase } from './lookup.js';

type MoveInput<T extends string> = {
  id: T;
  evaluationForMover: Evaluation;
};

export function classifyAnalyzedMoves<T extends string>(current: Evaluation, moves: Array<MoveInput<T>>): Array<ClassifiedMove<T>> {
  const bestIds = new Set(selectBestMoveIds(current, moves));
  return moves.map((move) => {
    const classification = classifyMove(current, move.evaluationForMover, bestIds.has(move.id));
    return {
      id: move.id,
      evaluationForMover: move.evaluationForMover,
      classification,
      plyDelta:
        current.result === move.evaluationForMover.result && current.ply !== undefined && move.evaluationForMover.ply !== undefined
          ? move.evaluationForMover.ply - current.ply
          : null,
      isBest: bestIds.has(move.id),
    };
  });
}

function selectBestMoveIds<T extends string>(current: Evaluation, moves: Array<MoveInput<T>>): T[] {
  if (moves.length === 0) return [];

  if (current.result === 'win') {
    const finite = moves.filter((move) => move.evaluationForMover.result === 'win' && move.evaluationForMover.ply !== undefined);
    if (finite.length === 0) return [];
    const minPly = Math.min(...finite.map((move) => move.evaluationForMover.ply ?? Number.POSITIVE_INFINITY));
    return finite.filter((move) => move.evaluationForMover.ply === minPly).map((move) => move.id);
  }

  if (current.result === 'draw') {
    const draws = moves.filter((move) => move.evaluationForMover.result === 'draw');
    if (draws.length > 0) return draws.map((move) => move.id);
    return moves.filter((move) => move.evaluationForMover.result === 'win').map((move) => move.id);
  }

  if (current.result === 'lose') {
    const improvements = moves.filter((move) => resultRank(move.evaluationForMover.result) > resultRank(current.result));
    if (improvements.length > 0) {
      const bestRank = Math.max(...improvements.map((move) => resultRank(move.evaluationForMover.result)));
      return improvements.filter((move) => resultRank(move.evaluationForMover.result) === bestRank).map((move) => move.id);
    }
    const finite = moves.filter((move) => move.evaluationForMover.result === 'lose' && move.evaluationForMover.ply !== undefined);
    if (finite.length === 0) return [];
    const maxPly = Math.max(...finite.map((move) => move.evaluationForMover.ply ?? Number.NEGATIVE_INFINITY));
    return finite.filter((move) => move.evaluationForMover.ply === maxPly).map((move) => move.id);
  }

  return [];
}

function classifyMove(current: Evaluation, evaluationForMover: Evaluation, isBest: boolean): MoveClassification {
  if (evaluationForMover.result === 'unknown' || current.result === 'unknown') return 'unknown';
  if (isBest) return 'best';
  if (current.result === 'win' && evaluationForMover.result === 'win') return 'inaccuracy';
  if (current.result === 'win' && evaluationForMover.result !== 'win') return 'blunder';
  if (current.result === 'draw' && evaluationForMover.result === 'lose') return 'mistake';
  if (current.result === evaluationForMover.result) return 'inaccuracy';
  return resultRank(evaluationForMover.result) < resultRank(current.result) ? 'mistake' : 'best';
}

function resultRank(result: Evaluation['result']): number {
  switch (result) {
    case 'win':
      return 3;
    case 'draw':
      return 2;
    case 'lose':
      return 1;
    default:
      return 0;
  }
}

export function invertEvaluation(evaluation: Evaluation): Evaluation {
  if (evaluation.result === 'win') return { result: 'lose', ply: evaluation.ply };
  if (evaluation.result === 'lose') return { result: 'win', ply: evaluation.ply };
  return { ...evaluation };
}

export function analyzeMoves(state: GameState, db: LookupDatabase) {
  const currentEvaluation = db.evaluateState(state);
  const generated = generateNextStates(state);
  const moveInputs = generated.map((move, index) => ({
    id: String(index),
    evaluationForMover: invertEvaluation(db.evaluateState(move.state)),
  }));
  const classified = classifyAnalyzedMoves(currentEvaluation, moveInputs);
  const moves = generated.map((move, index) => {
    const nextHash = HashCalc.hashToString(HashCalc.calcHash(move.state));
    const classification = classified[index];
    return {
      index,
      move: move.japanese,
      record: move.record,
      hash: nextHash,
      childEvaluation: db.evaluateState(move.state),
      evaluationForMover: classification.evaluationForMover,
      classification: classification.classification,
      classificationLabel: classificationLabel(classification.classification),
      plyDelta: classification.plyDelta,
      isBest: classification.isBest,
    };
  });

  return {
    hash: HashCalc.hashToString(HashCalc.calcHash(state)),
    turn: state.turn,
    currentEvaluation,
    moves,
    bestMoves: moves.filter((move) => move.isBest),
  };
}

export function getBestLine(state: GameState, db: LookupDatabase, maxPlies: number) {
  const steps: Array<{
    ply: number;
    hash: string;
    turn: GameState['turn'];
    bestMoves: ReturnType<typeof analyzeMoves>['bestMoves'];
  }> = [];
  let current = state;

  for (let ply = 0; ply < maxPlies; ply += 1) {
    const analyzed = analyzeMoves(current, db);
    steps.push({ ply, hash: analyzed.hash, turn: analyzed.turn, bestMoves: analyzed.bestMoves });
    if (analyzed.bestMoves.length === 0) return { steps, stoppedReason: 'no_legal_moves' as const };
    const selected = analyzed.bestMoves[0];
    const generated = generateNextStates(current)[selected.index] as GeneratedMove | undefined;
    if (!generated) return { steps, stoppedReason: 'internal_mismatch' as const };
    current = generated.state;
  }

  return { steps, stoppedReason: 'max_plies' as const };
}

export function classificationLabel(classification: MoveClassification): string {
  switch (classification) {
    case 'best':
      return '最善手';
    case 'inaccuracy':
      return '緩手';
    case 'blunder':
      return '敗着';
    case 'mistake':
      return '悪手';
    default:
      return '不明';
  }
}

File: src/mcp-go/analysis.go
package main

import "fmt"

type moveInput struct {
	id                 string
	evaluationForMover Evaluation
}

func classifyAnalyzedMoves(current Evaluation, moves []moveInput) []ClassifiedMove {
	bestIDs := selectBestMoveIDs(current, moves)
	bestSet := make(map[string]bool)
	for _, id := range bestIDs {
		bestSet[id] = true
	}

	result := make([]ClassifiedMove, len(moves))
	for i, move := range moves {
		classification := classifyMove(current, move.evaluationForMover, bestSet[move.id])
		var plyDelta *int
		if current.Result == move.evaluationForMover.Result && current.Ply != nil && move.evaluationForMover.Ply != nil {
			delta := *move.evaluationForMover.Ply - *current.Ply
			plyDelta = &delta
		}
		result[i] = ClassifiedMove{
			ID:                 move.id,
			EvaluationForMover: move.evaluationForMover,
			Classification:     classification,
			PlyDelta:           plyDelta,
			IsBest:             bestSet[move.id],
		}
	}
	return result
}

func selectBestMoveIDs(current Evaluation, moves []moveInput) []string {
	if len(moves) == 0 {
		return nil
	}

	if current.Result == EvalWin {
		var finite []moveInput
		for _, m := range moves {
			if m.evaluationForMover.Result == EvalWin && m.evaluationForMover.Ply != nil {
				finite = append(finite, m)
			}
		}
		if len(finite) == 0 {
			return nil
		}
		minPly := *finite[0].evaluationForMover.Ply
		for _, m := range finite[1:] {
			if *m.evaluationForMover.Ply < minPly {
				minPly = *m.evaluationForMover.Ply
			}
		}
		var result []string
		for _, m := range finite {
			if *m.evaluationForMover.Ply == minPly {
				result = append(result, m.id)
			}
		}
		return result
	}

	if current.Result == EvalDraw {
		var draws []string
		var wins []string
		for _, m := range moves {
			if m.evaluationForMover.Result == EvalDraw {
				draws = append(draws, m.id)
			}
			if m.evaluationForMover.Result == EvalWin {
				wins = append(wins, m.id)
			}
		}
		if len(draws) > 0 {
			return draws
		}
		return wins
	}

	if current.Result == EvalLose {
		var improvements []moveInput
		for _, m := range moves {
			if resultRank(m.evaluationForMover.Result) > resultRank(current.Result) {
				improvements = append(improvements, m)
			}
		}
		if len(improvements) > 0 {
			bestRank := resultRank(improvements[0].evaluationForMover.Result)
			for _, m := range improvements[1:] {
				r := resultRank(m.evaluationForMover.Result)
				if r > bestRank {
					bestRank = r
				}
			}
			var result []string
			for _, m := range improvements {
				if resultRank(m.evaluationForMover.Result) == bestRank {
					result = append(result, m.id)
				}
			}
			return result
		}

		var finite []moveInput
		for _, m := range moves {
			if m.evaluationForMover.Result == EvalLose && m.evaluationForMover.Ply != nil {
				finite = append(finite, m)
			}
		}
		if len(finite) == 0 {
			return nil
		}
		maxPly := *finite[0].evaluationForMover.Ply
		for _, m := range finite[1:] {
			if *m.evaluationForMover.Ply > maxPly {
				maxPly = *m.evaluationForMover.Ply
			}
		}
		var result []string
		for _, m := range finite {
			if *m.evaluationForMover.Ply == maxPly {
				result = append(result, m.id)
			}
		}
		return result
	}

	return nil
}

func classifyMove(current, eval Evaluation, isBest bool) MoveClassification {
	if eval.Result == EvalUnknown || current.Result == EvalUnknown {
		return ClassUnknown
	}
	if isBest {
		return ClassBest
	}
	if current.Result == EvalWin && eval.Result == EvalWin {
		return ClassInaccuracy
	}
	if current.Result == EvalWin && eval.Result != EvalWin {
		return ClassBlunder
	}
	if current.Result == EvalDraw && eval.Result == EvalLose {
		return ClassMistake
	}
	if current.Result == eval.Result {
		return ClassInaccuracy
	}
	if resultRank(eval.Result) < resultRank(current.Result) {
		return ClassMistake
	}
	return ClassBest
}

func resultRank(result EvaluationResult) int {
	switch result {
	case EvalWin:
		return 3
	case EvalDraw:
		return 2
	case EvalLose:
		return 1
	default:
		return 0
	}
}

func invertEvaluation(eval Evaluation) Evaluation {
	switch eval.Result {
	case EvalWin:
		return Evaluation{Result: EvalLose, Ply: eval.Ply}
	case EvalLose:
		return Evaluation{Result: EvalWin, Ply: eval.Ply}
	default:
		return eval
	}
}

type AnalyzeMovesResult struct {
	Hash              string                   `json:"hash"`
	Turn              string                   `json:"turn"`
	CurrentEvaluation Evaluation              `json:"currentEvaluation"`
	Moves             []AnalyzedMove           `json:"moves"`
	BestMoves         []AnalyzedMove           `json:"bestMoves"`
}

type AnalyzedMove struct {
	Index              int            `json:"index"`
	Move               string         `json:"move"`
	Record             MoveRecord      `json:"record"`
	Hash               string         `json:"hash"`
	ChildEvaluation    Evaluation     `json:"childEvaluation"`
	EvaluationForMover Evaluation     `json:"evaluationForMover"`
	Classification     MoveClassification `json:"classification"`
	ClassificationLabel string         `json:"classificationLabel"`
	PlyDelta           *int           `json:"plyDelta"`
	IsBest             bool           `json:"isBest"`
}

func analyzeMoves(state GameState, db *LookupDatabase) *AnalyzeMovesResult {
	currentEvaluation := db.EvaluateState(state)
	generated := generateNextStates(state)
	moveInputs := make([]moveInput, len(generated))
	for i, move := range generated {
		moveInputs[i] = moveInput{
			id:                 fmt.Sprintf("%d", i),
			evaluationForMover: invertEvaluation(db.EvaluateState(move.State)),
		}
	}
	classified := classifyAnalyzedMoves(currentEvaluation, moveInputs)

	moves := make([]AnalyzedMove, len(generated))
	for i, move := range generated {
		nextHash := HashCalc.HashToString(HashCalc.CalcHash(move.State))
		cl := classified[i]
		moves[i] = AnalyzedMove{
			Index:              i,
			Move:               move.Japanese,
			Record:             move.Record,
			Hash:               nextHash,
			ChildEvaluation:    db.EvaluateState(move.State),
			EvaluationForMover: cl.EvaluationForMover,
			Classification:     cl.Classification,
			ClassificationLabel: classificationLabel(cl.Classification),
			PlyDelta:           cl.PlyDelta,
			IsBest:             cl.IsBest,
		}
	}

	var bestMoves []AnalyzedMove
	for _, m := range moves {
		if m.IsBest {
			bestMoves = append(bestMoves, m)
		}
	}

	return &AnalyzeMovesResult{
		Hash:              HashCalc.HashToString(HashCalc.CalcHash(state)),
		Turn:              string(state.Turn),
		CurrentEvaluation: currentEvaluation,
		Moves:             moves,
		BestMoves:         bestMoves,
	}
}

type BestLineResult struct {
	Steps         []BestLineStep `json:"steps"`
	StoppedReason string         `json:"stoppedReason"`
}

type BestLineStep struct {
	Ply       int            `json:"ply"`
	Hash      string         `json:"hash"`
	Turn      string         `json:"turn"`
	BestMoves []AnalyzedMove `json:"bestMoves"`
}

func getBestLine(state GameState, db *LookupDatabase, maxPlies int) *BestLineResult {
	var steps []BestLineStep
	current := state

	for ply := 0; ply < maxPlies; ply++ {
		analyzed := analyzeMoves(current, db)
		step := BestLineStep{
			Ply:       ply,
			Hash:      analyzed.Hash,
			Turn:      analyzed.Turn,
			BestMoves: analyzed.BestMoves,
		}
		steps = append(steps, step)
		if len(analyzed.BestMoves) == 0 {
			return &BestLineResult{Steps: steps, StoppedReason: "no_legal_moves"}
		}
		selected := analyzed.BestMoves[0]
		generated := generateNextStates(current)
		if selected.Index >= len(generated) {
			return &BestLineResult{Steps: steps, StoppedReason: "internal_mismatch"}
		}
		current = generated[selected.Index].State
	}

	return &BestLineResult{Steps: steps, StoppedReason: "max_plies"}
}

func classificationLabel(classification MoveClassification) string {
	switch classification {
	case ClassBest:
		return "最善手"
	case ClassInaccuracy:
		return "緩手"
	case ClassBlunder:
		return "敗着"
	case ClassMistake:
		return "悪手"
	default:
		return "不明"
	}
}